### PR TITLE
RequestOptions for handling refresh tokens and http request failures

### PR DESCRIPTION
**Pull request by hpinkos**
Not ready to merge, looking for early feedback on this approach.

This solves the following problem:  Some data providers supply a refresh token that allows access to the tiles.  This token expires after a certain amount of time, and we currently don't have any way to update an expired token to a new once since we've always handled access tokens by just appending them to all URLs.

To fix this, I created a `RequestOptions` class with the following properties:
- `beforeRequest`: A callback function that takes a url and returns a url.  This can be used to append the refresh token onto the end of a base url
- `retryOnError`: A callback function that returns a promise to a boolean.  If it returns true, the http request is re-opened.  This function can be used to generate a new refresh token if the old one has expired.
- `retryAttempts`: The number of times a request should be retried.

In this branch, `createTileMapServiceImageryProvider`, `CesiumTerrainProvider` and `Cesium3DTileset` have been updated to use a `requestOption` parameter.  This is piped through to `loadXXX` (`loadJson`, `loadText`, etc).  `loadWithXhr`, `loadImage` and `loadJsonp` are the functions that actually make use of the `requestOption` parameter (our other load functions call these functions)

Using these options for `createTileMapServiceImageryProvider` would look something like this:

```js
var accessToken;
var requestOptions = new RequestOptions({
    retryAttempts: 1,
    beforeRequest: function(url) {
        return url + '?access_token=' + accessToken;
    },
    retryOnError = function() {
        return getNewAccessToken()
            .then(function(newToken) {
                accessToken = newToken;
                return true;
            })
            .otherwise(function() {
                return false;
            });
    }
});

return createTileMapServiceImageryProvider({
    url : metadata.url,
    requestOptions : requestOptions
});
```

Thoughts?  

@shunter do you have any feedback for this approach?

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Signed CLA is on file.

@hpinkos, thanks for the pull request! Maintainers, we have a signed CLA from @hpinkos, so you can review this at any time.

:warning: I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by emackey** [comment]
@swong-agi Not sure if this has relevance to your team, but thought I would ping you on this.

File: Source/Core/CesiumTerrainProvider.js [hunk 1]
```diff
@@ -284,7 +285,7 @@ define([
                 if (defined(that._proxy)) {
                     metadataUrl = that._proxy.getURL(metadataUrl);
                 }
-                var parentMetadata = loadJson(metadataUrl);
+                var parentMetadata = loadJson(metadataUrl, undefined, undefined, requestOptions);
```
Comment (mramato) [inline Source/Core/CesiumTerrainProvider.js]: You already know what I'm going to say: Forcing callers to pass in `undefined` as standard practice to skip arguments is bad API design.  Doesn't have to be this PR, but if this is going to go in largely as-is, we should submit an issue to update loadJson et al to take an options object sooner rather than later.

File: Source/Core/RequestOptions.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * A function that determines whether or not to retry a request
+     * @callback RequestOptions~RetryOnErrorCallback
+     * @returns {Promise<Boolean>} A promise for whether or not to retry the request
+     */
+
+    /**
+     * A function that is called before the request is opened.
+     * @callback RequestOptions~BeforeRequestCallback
+     * @params {String} url The url for the request
```
Comment (mramato) [inline Source/Core/RequestOptions.js]: When we talked offline, my thinking was that this callback would take the XHR so that other things can be manipulated and not just the url (headers for example).  But perhaps passing the entire XHR object is a leaky abstraction that should be avoided.  Still, I'm trying to think if there are other use cases that would be handled by this callback if we made other options available. Perhaps @shunter has some thoughts.

File: Source/Core/loadArrayBuffer.js [hunk 3]
```diff
@@ -28,12 +28,13 @@ define([
      * @see {@link http://www.w3.org/TR/cors/|Cross-Origin Resource Sharing}
      * @see {@link http://wiki.commonjs.org/wiki/Promises/A|CommonJS Promises/A}
      */
-    function loadArrayBuffer(url, headers, request) {
+    function loadArrayBuffer(url, headers, request, requestOptions) {
```
Comment (mramato) [inline Source/Core/loadArrayBuffer.js]: `request` is meant for internal use only, so we can't actually add new public options after it (so this might force us to switch to an options object sooner rather than later.

**Comment by mramato** [comment]
Just thinking outloud, but I was hoping `retryOnError` could take a status code, but I remember you saying that `loadImage` can't get to the status code.  Maybe we can just have loadImage have it be an undefined value and return the actual status from other stuff.

File: Source/Scene/UrlTemplateImageryProvider.js [hunk 4]
```diff
@@ -537,6 +538,24 @@ define([
...
+         * Gets the request options.  This function should not be called before {@link UrlTemplateImageryProvider#ready} returns true.
+         * @memberof UrlTemplateImageryProvider.prototype
+         * @type {Object}
+         * @readonly
+         * @default undefined
+         */
+        requestOptions : {
+            get : function() {
+                //>>includeStart('debug', pragmas.debug);
+                if (!this.ready) {
```
Comment (mramato) [inline Source/Scene/UrlTemplateImageryProvider.js]: Not sure this is needed for this property, since it's set in the constructor, right?

**Comment by mramato** [comment]
I think this implementation is definitely "a" solution to the problem and is entirely workable without upending Cesium's current approach to resource loading (which is why I like it), but I would love to hear others thoughts on improvements to this approach or different approaches entirely to solving this problem in a generic way for all resource loading. (Yes, @shunter I'm looking at you 😄 )

**Comment by shunter** [comment]
> workable without upending Cesium's current approach to resource loading

Perhaps it's time to actually upend the approach?  In this PR, `loadJson` now takes:

`url, headers, request, requestOptions`

It seems like this is 4 separate parameters to pass various semi-overlapping parts of the same conceptual thing - a remote resource which needs to be loaded.  When you consider `loadJson`, `loadArrayBuffer`, `loadXML` etc. etc. etc. then you see another piece of data - the type of the remote resource.

The actual goal of clients like `CesiumTerrainProvider` is to load sub-resources which are derived from a base resource.  Generally this could be done by creating a copy of the resource, then manipulating some parts of the definition, such as the URL, leaving other parts unchanged.  Then the copy is submitted to be requested, returning a promise for the result.

When you're not working with strings, then possibilities open up.  Rather than intercepting and rewriting url strings, the base resource could just handle refreshing and replacing the token internally.  As far as client code is concerned, it doesn't matter that the base resource appears to contain different tokens over time - the terrain provider is just going to clone the base request (snapshot) and then modify only the parts relevant (the path to the specific terrain tile needed).

Basically, we could create a single object called `Resource` or something similar, and use that exclusively. No more URL string concatenation (strings can be immediately converted to a simple Resource).  No more query string manipulation garbage (a Resource contains an actual params object with key-value pairs).  `joinUrls` moves into `Resource`.  `Request` moves into `Resource`.  `loadWithXhr` takes a `Resource` only.

**Comment by mramato** [comment]
I'm not against the Resource idea, since we can always allow a string.  My main concern is what does that object look like?  It obviously needs to identify the primary url (for example layer.json or a czml file), but how exactly is it used to create derived urls? @shunter do you have a few minutes to talk offline with me and @hpinkos ?  I'm not against upending the status quo, but it would be faster to discuss some of this in person.

The nice thing about the Resource object is that it will also fix some of the other open issues we have with Model entity headers and at least one other one that escapes my mind at this moment.

**Comment by shunter** [comment]
Also consider the differences in how images are requested by imagery providers - `loadImage` won't be able to handle request headers, for example.  Automatic transparent fallback to `loadImageViaBlob` can be done if the request needs it (for performance reasons, it should prefer direct `loadImage` if it will work for a given resource).

See discussion thread: https://groups.google.com/d/topic/cesium-dev/3NxyEFmMzDM/discussion

**Comment by hpinkos** [comment]
Closing in favor of the approach in #6035

**Comment by cesium-concierge** [comment]
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/d/topic/cesium-dev/3NxyEFmMzDM/discussion

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/pull/6019 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.
---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:
